### PR TITLE
BUG-16807-1 SparseFrame fills with default_fill_value if data is None

### LINKED ISSUE #16807: BUG: SparseDataFrame constructor issues
I've noticed a bunch of issues with SparseDataFrame's constructor, particularly around support for non-`float64` datatypes:

- [ ]  when `data` is not provided, the default_fill_value is not used to fill the data (noted in passing #5470):
```python
>>> pd.SparseDataFrame(columns=list("ab"), index=range(4), default_fill_value=0.0)
    a   b
0 NaN NaN
1 NaN NaN
2 NaN NaN
3 NaN NaN
```
~~- [ ] when `data` is not provided and the `dtype` is set to `int64`, the constructor call fails~~
```python
>>> pd.SparseDataFrame(columns=list("ab"), index=range(4), dtype=np.int64)
ValueError: cannot convert float NaN to integer
```

- [ ] ~~when `data` is an empty sparse matrix with `int64` entries and no `dtype` is specified, the type of the SparseDataFrame is still `float64`. Worse, even when `dtype` is specified, it still doesn't work.~~
```python
>>> pd.SparseDataFrame(coo_matrix((4,3), dtype=np.int64))
    0   1   2
0 NaN NaN NaN
1 NaN NaN NaN
2 NaN NaN NaN
3 NaN NaN NaN
>>> pd.SparseDataFrame(coo_matrix((4,3), dtype=np.int64), dtype=np.int64)
    0   1   2
0 NaN NaN NaN
1 NaN NaN NaN
2 NaN NaN NaN
3 NaN NaN NaN
```
- [ ] ~~probably due to the same issue, given a sparse matrix with underlying elements of type `int64`, if any of the values in a column is `0` then that column will be treated as `float64`s:~~
```python
>>> pd.SparseDataFrame(coo_matrix(np.arange(12).reshape(4,3), dtype=np.int64))
     0   1   2
0  NaN   1   2
1  3.0   4   5
2  6.0   7   8
3  9.0  10  11
```
(as long as at least one value is non-zero, then this can be worked around by setting `default_fill_value`; however, if all of the values in one column are zero then this no longer works).

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.2
pytest: 2.9.2
pip: 8.1.2
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.13.0
scipy: 0.18.1
xarray: None
IPython: 5.1.0
sphinx: 1.4.6
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: 1.1.0
tables: 3.2.2
numexpr: 2.6.1
feather: None
matplotlib: 2.0.2
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_gbq: None
pandas_datareader: 0.4.0

</details>

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [ ] closes #16807
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Previously, the constructor for SparseFrame would default to `np.nan` if the data parameter was `None`, even if a `default_fill_value` is specified. This PR fixes this.

### CODE REVIEW

**Review by gfyoung — APPROVED** [review]
cc @jreback

**Comment by jreback** [comment]
there are a number of issues in the OP, are these all tested?

**Comment by JustinZhengBC** [comment]
@jreback I added tests for issues 2-3. The last one I think isn't a storage issue, but a formatting one. Sometimes, integers in sparse data structures can be printed as floats if there are `NaN` values involved. Can it be fixed in this PR or should it be considered a separate issue?
```
>>> a=pd.SparseDataFrame(coo_matrix(np.arange(2).reshape(2,1),dtype=np.int64))
>>> a
     0
0  NaN
1  1.0
>>> a[0][1]
1
```

**Review by WillAyd — APPROVED** [review]
lgtm - can you merge master and ping on green?

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1835,6 +1835,7 @@ Sparse
 - Bug in :meth:`SparseArray.nonzero` and :meth:`SparseDataFrame.dropna` returning shifted/incorrect results (:issue:`21172`)
 - Bug in :meth:`DataFrame.apply` where dtypes would lose sparseness (:issue:`23744`)
 - Bug in :func:`concat` when concatenating a list of :class:`Series` with all-sparse values changing the ``fill_value`` and converting to a dense Series (:issue:`24371`)
+- Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.rst]: Move to 0.25

**Review by WillAyd — CHANGES_REQUESTED** [review]
Whoops sorry one comment on whatsnew

File: doc/source/whatsnew/v0.24.2.rst [hunk 2]
```diff
@@ -97,6 +97,7 @@ Bug Fixes
 
 - Bug in :meth:`Series.is_unique` where single occurrences of ``NaN`` were not considered unique (:issue:`25180`)
 - Bug in :func:`merge` when merging an empty ``DataFrame`` with an ``Int64`` column or a non-empty ``DataFrame`` with an ``Int64`` column that is all ``NaN`` (:issue:`25183`)
+- Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.2.rst]: can you move to 0.25

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. merge master; ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by JustinZhengBC** [comment]
@jreback green

**Comment by gfyoung** [comment]
Thanks @JustinZhengBC !

**Pull request merged**